numbat-cli/src/main.rs

```diff
@@ -353,121 +355,134 @@ impl Cli {
     ) -> Result<()> {
         let mut session_history = SessionHistory::default();
 
+        let cmd_runner = CommandRunner::<Editor<NumbatHelper, DefaultHistory>>::new_all_disabled()
+            .enable_help(|| {
+                let help = help_markup();
+                print!("{}", ansi_format(&help, true));
+                // currently, the ansi formatter adds indents
+                // _after_ each newline and so we need to manually
+                // add an extra blank line to absorb this indent
+                println!();
+                CommandControlFlow::Normal
+            })
+            .enable_info(|ctx, item| {
+                let help = ctx.print_info_for_keyword(item);
+                println!("{}", ansi_format(&help, true));
+                CommandControlFlow::Normal
+            })
+            .enable_list(|ctx, item| {
+                let m = match item {
+                    None => ctx.print_environment(),
+                    Some(command::ListItems::Functions) => ctx.print_functions(),
+                    Some(command::ListItems::Dimensions) => ctx.print_dimensions(),
+                    Some(command::ListItems::Variables) => ctx.print_variables(),
+                    Some(command::ListItems::Units) => ctx.print_units(),
+                };
```

Comment:
This looks like something that could be part of the command runner? The callback would then directly take the `Markup` as an argument?

We could go one step further and let frontends basically specify a Markup=>Output function once. And that could be re-used generically for `help`, `info`, `list …`, etc.

Reply:
Just to clarify: you're saying that the user of the command runner would not provide the whole closure above, just `println!("{}", ansi_format(&m, false));`, and the command runner itself would run that match statement and return `Continue` (since there is no other reasonable way to implement the guts of `list`)?

Reply:
Exactly

Reply:
I think basically fits the bill now.